Add validation to EDI

reggieriser opened the pull request:
## Description

This PR adds validation to the invoicing EDI generation process to avoid invalid EDI being sent to GEX.

Given the amount of structs, validations, and nesting involved, I am leveraging a third-party [`validator` package](https://github.com/go-playground/validator).  This package allows us to use struct tags to define validators, thus making the validations obvious when looking at a struct's fields.  It also provides a lot of built-in and tested validations out of the box, and can handle validation of nested structs and slices of structs.

## Reviewer Notes

Given that we have mothballed much of the HHG code, it is no longer possible to test this via the UI.  ~~For now, we do still have EDI-related tests that are run as part of `make server_test`.  I have also created a new `TestValidate` test to spot check validations.~~

~~For security reasons, I do not return explicit validation errors via the endpoint.  Instead, I return a generic message and log the validation errors.~~

Update: Further HHG mothballing got rid of the shipment-centric EDI code/tests I was using previously in this PR, so I have shifted to tests around the individual EDI segments (which still remain and may be used again for future invoicing).

Note that the validations on the struct fields are sometimes more restrictive than the EDI documents allow -- this was tailored to the specific use cases we had before.  We can always relax these as we begin to use these structs again in the future.

## Setup

`make clean server_test`

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/tree/master/docs/backend.md#logging)
* [ ] Request review from a member of a different team.
* [ ] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/166531456) for this change

reggieriser commented:
I'm marking this one as a work-in-progress because the HHG mothballing work got rid of the shipment-based invoicing code/tests which I was relying on.  Instead, I'm working on EDI segment-based tests since those segments remain.

reggieriser commented:
I'm removing the WIP tag since I have merged all the master changes (and the deletions that affected this PR) and added in EDI segment-based tests.  It's ready for review.

reggieriser commented on pkg/services/tsp/tsp_service_test.go:
```diff
@@ -18,7 +18,7 @@ func (suite *TSPServiceSuite) SetupTest() {
 }
 
-func TestUserSuite(t *testing.T) {
+func TestTSPServiceSuite(t *testing.T) {
```
This isn't directly related to this PR, but I just happened to notice this naming issue (which I previously introduced) as I was looking for examples of testing suites.

kimallen commented on pkg/edi/segment/bx_test.go:
```diff
@@ -0,0 +1,65 @@
+			TransactionSetPurposeCode:    "01",      // eq
+			TransactionMethodTypeCode:    "K",       // eq
+			ShipmentMethodOfPayment:      "QQ",      // eq
+			ShipmentIdentificationNumber: "A-12345", // alphanum
```
I thought that `-` counted as an alphanumeric character- is that incorrect? what is it that makes this one fail?

reggieriser replied:
For the purposes of this `validator` library, `alphanum` is defined as just A-Z (upper or lower case) and 0-9.  [Here's the regex](https://github.com/go-playground/validator/blob/556b9da3c05f2a935bc086fb5a0bb55dd8112d2d/regexes.go#L7) they're using.

Seems like it's a little open to interpretation [per Wikipedia](https://en.wikipedia.org/wiki/Alphanumeric).

kimallen replied:
I'm pretty sure the ASC X12 allows special characters.  It at least allows spaces.  Although it's for rail specs, this [source](http://www.railcis.org/information-standards/edi-guidelines/edi-introduction-and-overview/asc-x12-standards/) says `AN String Values for the string-type data elements are a sequence of any letters, digits, spaces, and/or special characters.`[This one](http://www.edibasics.co.uk/wp-content/uploads/Tutor_ANSI.SO_GXS.pdf) on p. 15 does not specify about special symbols, but does say spaces are okay.  We may need to write a custom validator that includes special chars.  I could email GEX folks and they may be able to tell me for sure.  Let me know if you want me to do that.

kimallen replied:
I'm seeing in the spec for BX04 that this one specifically does not include spaces or special characters, so this one is okay to use `alphanum` as the validator.  I wonder if there are places that ask for AN that don't specify, so do allow spaces and special characters.  I'll look as I walk through.

reggieriser replied:
Right, I did notice that difference between the `alphanum` validator and the `AN` for EDI (and avoided `alphanum` in most places because of that).  However, there were a few places like this one where `alphanum` was actually appropriate.  Always possible I missed something, though.

kimallen commented on pkg/edi/segment/isa.go:
```diff
@@ -7,22 +7,22 @@ import (
+	InterchangeControlNumber          int64  `validate:"min=1,max=999999999"`
+	AcknowledgementRequested          int    `validate:"eq=1"`
+	UsageIndicator                    string `validate:"oneof=P T"`
+	ComponentElementSeparator         string `validate:"eq=0x7C"` // Have to escape pipe symbol
```
I don't think it matters what we use, but I'm wondering why we are using pipe now, whereas we were using a `*` before

kimallen replied:
Never mind- I think I was thinking of the "DataElementSeparator", which is different

kimallen commented on pkg/edi/segment/bx.go:
```diff
@@ -6,13 +6,13 @@ import (
+	TransactionSetPurposeCode    string `validate:"eq=00"`
+	TransactionMethodTypeCode    string `validate:"eq=J"`
+	ShipmentMethodOfPayment      string `validate:"eq=PP"`
+	ShipmentIdentificationNumber string `validate:"alphanum,min=1,max=30"`
```
I'm trying to find the spec but not finding it- how do we know the value cannot be 0?

reggieriser replied:
For `BX04`, the spec does show `O` (optional) and `AN 1/30` (alphanumeric between 1 and 30 length), but then also says that the GBL Number should always be included and should not use any punctuation or special characters.  So, I assumed that meant for our purposes (at least our previous use case) it should not allow an empty string.  (Note that you mentioned a value of "0" -- that would actually satisfy a `min=1` per the validator since this is a `string` and not an `int`.)

kimallen commented on pkg/edi/segment/mea.go:
```diff
@@ -7,9 +7,9 @@ import (
-	MeasurementValue           float64
+	MeasurementReferenceIDCode string  `validate:"omitempty,len=2"`
+	MeasurementQualifier       string  `validate:"omitempty,oneof=HT LN WD"`
+	MeasurementValue           float64 `validate:"required"`
```
Does this one need a min/max?

reggieriser replied:
Doing that is tricky because this is a `float64`.  The EDI spec says `1/20` (`The length of a decimal type data element does not include the optional minus signs, decimal point, or trailing exponent indicator “E.”`).  A `min` or `max` in the validator is testing something different -- it's whether the `MeasurementValue` is between the `min` and `max` values inclusive (as opposed to testing length).  I think we'd have to write a custom validator in this case, and that is difficult given a `float64` is just an approximation.

If we wanted to enforce length with struct field validations, we'd probably have to refactor the float64 fields to another data type like a `string` or `int` (storing in cents, perhaps) that's more precise.  That seems outside the scope of this story, though.

kimallen replied:
This seems sufficient then, thanks for looking into it

reggieriser commented:
Thanks for the detailed review!  I've addressed the comments -- let me know what you think.

As far as the `LX` segment, I couldn't find it referenced in the docs anywhere (nor the `Assigned Number` field).  We also didn't seem to use `LX` in our previous invoicing work.  So, I left it validation-less for now since I don't know the rules for it (although it is already constrained some by being an `int` type).

reggieriser commented:
@jacquelineIO ... yes, you can add custom validations either at the field level or the overall struct level.  The `timeformat` validator is one I wrote -- see its definition [here](https://github.com/transcom/mymove/blob/rr-validate-edi/pkg/edi/segment/validators.go) and its registration [here](https://github.com/transcom/mymove/blob/9dc61b24f2ea339eaeb1ebb62d7b8e48963d7a67/pkg/edi/segment/segment_test.go#L24).

There are other examples linked [in the validator library docs](https://github.com/go-playground/validator#usage-and-documentation).

jacquelineIO commented:
> @jacquelineIO ... yes, you can add custom validations either at the field level or the overall struct level. The `timeformat` validator is one I wrote -- see its definition [here](https://github.com/transcom/mymove/blob/rr-validate-edi/pkg/edi/segment/validators.go) and its registration [here](https://github.com/transcom/mymove/blob/9dc61b24f2ea339eaeb1ebb62d7b8e48963d7a67/pkg/edi/segment/segment_test.go#L24).
> 
> There are other examples linked [in the validator library docs](https://github.com/go-playground/validator#usage-and-documentation).

nice! I don't have any further comments, approved from my view. I see Kim has it covered as well!

kimallen commented:
> As far as the `LX` segment, I couldn't find it referenced in the docs anywhere (nor the `Assigned Number` field). We also didn't seem to use `LX` in our previous invoicing work. So, I left it validation-less for now since I don't know the rules for it (although it is already constrained some by being an `int` type).

I did find it in [this doc](https://www.ustranscom.mil/cmd/associated/dteb/files/transportationics/dteb220b004010_Version_0.pdf) on page 19, but if we don't actually use it, no need to write the validation (I was just walking down all the segments in the pkg)

reggieriser commented:
> I did find it in [this doc](https://www.ustranscom.mil/cmd/associated/dteb/files/transportationics/dteb220b004010_Version_0.pdf) on page 19, but if we don't actually use it, no need to write the validation (I was just walking down all the segments in the pkg)

I must have missed that doc.  For completeness, I went ahead and added a validation and test for `LX`.
